lib/model/message.dart

```diff
@@ -409,20 +409,76 @@ class MessageStoreImpl extends HasChannelStore with MessageStore, _OutboxMessage
     // from the event queue, so there's inherently a race.
     //
     // If the fetched message reflects changes we haven't yet heard from the
-    // event queue, then it doesn't much matter which version we use: we'll
-    // soon get the corresponding events and apply the changes anyway.
-    // But if it lacks changes we've already heard from the event queue, then
-    // we won't hear those events again; the only way to wind up with an
+    // event queue, then normally (see [1]) it doesn't much matter which version
+    // we use: we'll soon get the corresponding events and apply the changes
+    // anyway. But if it lacks changes we've already heard from the event queue,
+    // then we won't hear those events again; the only way to wind up with an
     // updated message is to use the version we have, that already reflects
-    // those events' changes.  So we always stick with the version we have.
+    // those events' changes.  So we always stick with the version we have. [1]
+    //
+    // [1] With one exception: if the version we have was in an unsubscribed
+    //     channel when we got it or sometime since, we take the fetched version
+    //     instead. That's because our version might be stale; we don't expect
+    //     update events for messages in unsubscribed channels.
     for (int i = 0; i < messages.length; i++) {
       final message = messages[i];
-      messages[i] = this.messages.putIfAbsent(message.id, () {
-        message.matchContent = null;
-        message.matchTopic = null;
-        return message;
-      });
+
+      messages[i] = this.messages.update(message.id,
+        (current) => _reconcileWhenPresent(current, message),
+        ifAbsent: () => _reconcileWhenAbsent(message));
+    }
+  }
+
+  Message _reconcileWhenPresent(Message current, Message incoming) {
+    bool currentIsMaybeStale = false;
+    if (incoming is StreamMessage) {
+      if (subscriptions[incoming.streamId] != null) {
+        // The incoming message won't grow stale; it's in a subscribed channel.
+        // Remove it from _maybeStaleChannelMessages if it was there.
+        currentIsMaybeStale = _maybeStaleChannelMessages.remove(incoming.id);
+      } else {
+        assert(_maybeStaleChannelMessages.contains(incoming.id));
+        currentIsMaybeStale = true;
+      }
+    }
+
+    return currentIsMaybeStale
+      ? _stripMatchFields(incoming)
+      : current;
```

Comment:
I think this logic would benefit from having the explanation in the long comment above worked into it.

Probably that mostly means just moving that comment from `reconcileMessages` into this method. After all, it starts with:
> What to do when some of the just-fetched messages are already known?

so it's entirely about this "present" case.

Beyond that, I think once the comment is here in this method, the "normally" vs. "with one exception" cases can be worked a bit into the corresponding cases in this code's control flow.